src/gf.c

```diff
@@ -611,9 +613,9 @@ static void jl_cacheable_sig(
         }
         else if (notcalled_func && (decl_i == (jl_value_t*)jl_any_type ||
                                     decl_i == (jl_value_t*)jl_function_type ||
-                                    (jl_is_uniontype(decl_i) && jl_svec_len(((jl_uniontype_t*)decl_i)->types)==2 &&
-                                     jl_subtype((jl_value_t*)jl_function_type, decl_i, 0) &&
-                                     jl_subtype((jl_value_t*)jl_datatype_type, decl_i, 0)))) {
+                                    (jl_is_uniontype(decl_i) && //jl_svec_len(((jl_uniontype_t*)decl_i)->types)==2 &&
```

Comment:
if we want to keep this as a test for equivalency to `Callable`, the equivalent here would probably be `jl_is_datatype(((jl_uniontype_t*)decl_i)->b)`. Or could directly test `(u.a == jl_function_type && u.b == jl_datatype_type) || (u.b == jl_function_type && u.a == jl_datatype_type)`